### ISSUE TITLE
[OJS] Make notification emails for new announcements and issues optional in OJS3

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2561]
Hi @NateWr, I was a "bit" slow with this...

Issue described here: http://forum.pkp.sfu.ca/t/make-notification-emails-for-new-announcements-and-issues-optional-in-ojs3-0-2/29513

Basically a lot of journals would like to disable the notification sent to all registered users when an annoucement is written or a new issues is published. Also, the very short notification messages usually cause questions among the users.

Three options with this one:
1. Journal settings for enabling/disabling the notifications for new annoucements and issues
2. No settings, but a unsbuscribe button: "I do not want to receive these kinds of notifications anymore" (see @ctgraham's comments in the forum thread)
3. A combination of the two options

Number 1 would be fairly easy to accomplish, the biggest question is where to put the setting selections. Number 2 is clearly a lot of work, but would be a very nice if combined with the settings you already have in user profile.

**Comment by NateWr** [comment]
I flagged this as a Major Feature for the first part, to disable emails on announcements and new issues. There seems to be a lot of requests for this.

I think the other work, looking at how users can manage their email notifications, would align with other work we want to do on pinning down the notifications used throughout the system.

**Comment by ajnyga** [comment]
Hi @NateWr 

I could do a pr if you decide where the options should be. As I said in the original forum post, there are three options (in my opinion)
1. A separate page under workflow
2. In workflow => emails
3. The other one in Annoucements tab and the other one under Issues

**Comment by NateWr** [comment]
I'd like to get some feedback from @asmecher about the best way to handle this. I'm not very familiar with the email notification system.

Unfortunately, we're a bit busy right now prioritizing two big merges -- versioning and the JS changes. So I don't know if he'll have time to put more thought into this until closer to OJS 3.1's release.

**Comment by ajnyga** [comment]
Hi,

Controlling the notifications in both cases is fairly straight forward. There is already a rule in the new issue notifications that they are not sent if OJS is not used for publishing content (just the editorial process). So extending that is not a problem.

No rush, I am going to for a vacation starting form tomorrow, so I would not have time to look at this until around the 20th.

**Comment by asmecher** [comment]
In OJS2, we had a manual "Notify Users" feature that I think Editors appreciated and used heavily. We don't have that yet in OJS3, and I think most users find the automatic notification annoying and unpredictable. I'd favour removing the automatic notification entirely and rewriting the manual notification.

**Comment by ajnyga** [comment]
I think that the notify users feature was only related to the publishing of new issues? I could be wrong. 

Maybe we could have:
- title field
- pull down menu for prefilling the notification content (like in add participant feature)
- free text field
- checkboxes for choosing which roles receive the notification (these could be hidden by default)

Would this be under Users&Roles?

**Comment by asmecher** [comment]
> Would this be under Users&Roles?

I was picturing this under Tools, but I don't have a strong opinion on this.

> I think that the notify users feature was only related to the publishing of new issues? I could be wrong.

Yes, that's correct. For announcements, I was imagining simply a "Send Notification" checkbox on the announcement create/edit form. (But as above, I haven't thought about this in depth.)

**Comment by ajnyga** [comment]
Thanks, it's your call. Both solve the original issue.

Do you see problems in having a multipurpose notification tool?  I started to think that this could be very handy for journals, that is  if they do not use it too much.

**Comment by asmecher** [comment]
@ajnyga, I do think a multi-purpose tool will be valuable. There's potentially some overlap, too, with the Discussions tool -- I think we may need to eventually add some templating support, so that commonly-used boilerplate can be easily accessed.

The current notifications system is mostly a holdover from OJS 2.x, and never felt smoothly integrated, either by user experience or coding. I think we should be open to major changes there. @NateWr, pie-in-the-sky user-oriented thinking is welcome here.

**Comment by NateWr** [comment]
When we talk about a "multi-purpose tool", are we just talking about something to email all or a segment of users? Or are we talking about addressing the whole notifications system (batch emails, workflow emails, tasks)?

**Comment by asmecher** [comment]
The existing notification system does a lot of different things:

 - Instant notifications ("your form was saved")
 - Persistent notifications and tasks ("all reviews are in, make a decision")
 - Emailed notifications

There's room for improvement in all three, but I suspect the emailed notifications are by far the weakest link. We tried to design it so that overlap between them was shared (e.g. a task would also send an email) but I'm not convinced that it was worth binding these together.

**Comment by NateWr** [comment]
I agree, and I think the email-all-users-on-new-issues/announcements is a separate issue altogether. I'd recommend just addressing the bulk emailing aspect in this issue.

I've done some work in trying to handle status issues in the submissions list work. I'd like to continue in that vein for the other notification use-cases.

**Comment by ajnyga** [comment]
Hi,
How about 
1. adding an on/off setting to the new issue notification.
2. and then changing the announcement function so that you can choose if you want to send the new announcement you are writing to users (+ choose which roles as well). 

This way the announcement function would work also as the "multi-purpose email tool".

As a separate but related issue, I think it would be important that in the case of announcements the actual title and content would be in the notification email. At the moment the email just says something like "a new announcement was added" which does not serve the users well (a whole different thing are the notifications related to the editorial process, where you do not want to receive replies to the email itself).

**Comment by jmacgreg** [comment]
Hey @ajnyga we've received reports from clients being surprised by the frequency of these two notification types (new content published; announcement notifications). I think your solution here sounds great: it allows for folks to opt in/out of notifications, and also makes the Announcements function a more useful general tool. So +1 from me. I'd be happy to help test anything you develop, if you like.

**Comment by ajnyga** [comment]
Thanks @jmacgreg. I can do a pr on this issue, but let's see what @asmecher and @NateWr decide. I like the idea of a general email tool, but just the ability to control these two notifications is fine too.

**Comment by NateWr** [comment]
1. I'm in favor of adding settings to turn issue and announcement notifications off at the journal level (one setting for each).
2. I think that @ctgraham's comment about user unsubscribing is also just as important. A user should be able to unsubscribe from issue and announcement notifications.
3. I'm conflicted about turning announcements into a general email blasting tool. I like that it provides a feature with very little technical investment. However, I worry about the general consequences of trying to shoehorn one activity into a tool built for another (this is part of the problem we have with the internal notifications/tasks/alerts system), as well as the hazards of making it easier to send out email blasts.

I understand that email alerts and lists are the norm in the academic community (although every academic I know is drowning in their inbox so I'm not sure it's serving them well). Looking at emailing services, there's usually a whole bunch of parallel features that go into managing email lists: email previews, scheduled sending, results reporting, bounce tracking, unsubscribe flows, spam flag tracking, etc.

These things help people send better email by making the sender take it more seriously. Mailchimp's [big red button](http://www.gamificationgeek.com/updates/2016/6/27/mailchimp-and-the-big-red-button) is a good example. There's a whole ceremony to the process which kind of clarifies the enormity of the act in the sender's mind.

So I'm hesitant to slide in a feature like this as a kind of off-shoot of an existing feature. Though, fair enough, maybe I'm just bikeshedding.

Other issues to consider:
- Will email-only announcements have an Announcement Type? What would someone understand that to mean in that context?
- I think we'll at the very least need a confirmation step: "You're about to send an email to 1,231 recipients. Are you sure?" This might be good to add to issues and announcements anyway when the emailing is enabled.

**Comment by Ph-We** [comment]
Many of our journal managers ask for a somewhat contrary option: they need a simple form (email, name) for users to subscribe to notifications for new announcements and issues.

**Comment by t4x0n** [comment]
Hi @NateWr, I'm Editor in three journals and have some knowledge about code, so I'm talking from two sides of the problem.  

I don't know how and how much you are work in this issue, but I have an idea for implementing an easy way to stop sending short notifications for announcements and when a new issue is published (in this case we have problems too because we make some proof before the final publication of the issue, and users receive a lot of notifications): I think in a checkbox, like when you are in the revision workflow, and you can choose for not send an email to the author:

![image](https://user-images.githubusercontent.com/5422984/29487693-37cd6ba2-84d4-11e7-9b92-50c23903bca6.png)

this way is a probably faster implementation for announcements notifications and new issue notifications, and permit you to work independently in a solution for unsubscribing users.

My idea is similar to @ajnyga but maybe not identical, I just try to help. 

Thanks to all of you for your work!! 

Sorry for my English, my native language is Spanish.

**Comment by t4x0n** [comment]
Hi @NateWr, 

thanks for your answer. I am according with you about the human error, so yes, is good to have an option for control a default behavior and avoid sending emails by mistake.

**Comment by ajnyga** [comment]
@NateWr  I would be happy to help here with a pr here, but I am unsure on how you want to solve this?

If we just add the settings for turning on/off the notifications for new issues and announcements, then this will be a very fast addition. The only question is, that where exactly do we put these selections? For the announcements it is quite clear, but how about the issue notifications?

**Comment by NateWr** [comment]
@ajnyga I've just reviewed some of the discussion. It doesn't sound like we've come to any hard conclusions, but there does seem to be a consensus that options for disabling are important. My preference would be:

1. Journal-level settings for sending/not-sending email with an announcement and a new issue. These should act as defaults.
2. When sending an announcement, by default the send/not-send checkbox is based on the journal setting. This can be overridden for each announcement.
3. Issues could also have an override when choosing to publish an issue. I could take or leave this though.
4. User profile forms should have an option to choose to not receive announcement and new issue notification emails on a per-user basis.

I'd probably want to hear from @jmacgreg on how it fits the requests he's heard. And you'd probably want @asmecher to sign off before starting on anything, but he may be busy in the next few weeks as we're getting ready for 3.1 to go out.

**Comment by ajnyga** [comment]
Thanks! I could cover 1-3 fairly fast.  However, with 4 I am not sure. I am not familiar with the code there.

**Comment by asmecher** [comment]
@ajnyga and @NateWr, sorry for the wait. @NateWr, I agree with your recommendations on [this comment](https://github.com/pkp/pkp-lib/issues/2561#issuecomment-328144484), with a few minor caveats:

1/2. I'd prefer not to add journal-level settings for defaults -- I can see the logic in what you propose, but in the interests of combating settings form bloat, I'd suggest just fixing a default state on the annoucement form and leaving it up to the JM to uncheck it if they don't want to send anything. Suggest "send the announcement" as a default state for new announcements, and "don't send the announcement" as a default for editing existing announcements.
4. There's already a notification settings form available to each user on Profile > Notifications. It's not the crispest implementation we've ever done but extending this with the same pattern to include Announcements and New Issues should be straight-forward.

**Comment by ajnyga** [comment]
Thanks @asmecher! No problem, I realize that this is not the only open issue here.

If we would just add the profile setting for these, then I think that the problem would not be solved in the extend many editors would hope. Many journals have hundreds of users and I would say that at least about 80% of those have not visited the journal site in the recent years or are familiar with OJS3. 

So if the only way of opting out of these notification emails is editing the user profile, most of the users do not know how to do that and/or do not want to find out. What they may do is to reply to the email they received and say please stop :)

The other problem for leaving out journal defaults is the situation where the journal is adding old issues. In these cases the journal probably never wants to send a notification email, because we are not dealing with an actual new issue, but at the same time may want to send one when publishing an actual new issue.

So my suggestion is:
- No journal settings for these two notifications
- When publishing an issue a modal window will open where you can check/uncheck if you want to send the notification by email
- When publishing/editing an announcement add an extra checkbox to control the notification email (+ add the actual text of the announcement to the email, ok?)
- add profile settings to disable/enable the two notification types (AND add unsubscribe link to the notification emails which enables the user to disable the type of notification with one click, this could be added later on, but it would fit the overall design there very well)

If this sounds good, I can start working on it.

**Comment by asmecher** [comment]
Absolutely! This sounds perfect.

**Comment by ajnyga** [comment]
Hi @asmecher and @NateWr 
These two pr's solve 1-3 but without the journal defaults as discussed above
Announcements: https://github.com/pkp/pkp-lib/pull/2804
Publish issues: https://github.com/pkp/ojs/pull/1542

**Comment by asmecher** [comment]
Wow, @ajnyga, that's probably the quickest community contribution I've ever seen.

I think this deals with 90% of the frustration over this issue -- I've filed the remaining piece as https://github.com/pkp/pkp-lib/issues/2805 for tracking purposes, in case it doesn't make the threshold for OJS 3.1. Huge thanks as always, @ajnyga.

**Comment by ajnyga** [comment]
Great, thanks!

**Comment by amirtkh** [comment]
I installed the Open Journal System (OJS) version of 3.1.1.4 to handle our journals. A few weeks ago notification sending by OJS was suddenly stopped. Based on my searches on Google it is a prevalent problem in OJS and many users have been faced with it. Also, there is no clear solution to solve this problem. We edited the config.inc.php file as follows, but the OJS notifications did not work. I show the edited codes in the config.inc.php file for you. 

; Use SMTP for sending mail instead of mail()
; smtp = On

; SMTP server settings
; smtp_server = mail.********.com
; smtp_port = 25

; Enable SMTP authentication
; Supported mechanisms: ssl, tls
; smtp_auth = ssl
; smtp_username = ******@******.com
; smtp_password = *******

I could find another solution to solve the problem. It was editing the PHPMailer.php file to introduce e-mail parameters to OJS. I looked for the PHPMailer.php file on the server but I could not find it on the public_html/lib/pkp/lib/vendor/phpmailer/phpmailer/src/PHPMailer.php. This section of address "/src/PHPMailer.php" did not exist on the server! Is there any alternative solution to solve this problem?

**Comment by ctgraham** [comment]
> I installed the Open Journal System (OJS) version of 3.1.1.4 to handle our journals. A few weeks ago notification sending by OJS was suddenly stopped. Based on my searches on Google it is a prevalent problem in OJS and many users have been faced with it. Also, there is no clear solution to solve this problem. We edited the config.inc.php file as follows, but the OJS notifications did not work. I show the edited codes in the config.inc.php file for you.
> 
> ; Use SMTP for sending mail instead of mail()
> ; smtp = On
> 
> ; SMTP server settings
> ; smtp_server = mail.********.com
> ; smtp_port = 25
> 
> ; Enable SMTP authentication
> ; Supported mechanisms: ssl, tls
> ; smtp_auth = ssl
> ; smtp_username = ******@******.com
> ; smtp_password = *******
> 
> I could find another solution to solve the problem. It was editing the PHPMailer.php file to introduce e-mail parameters to OJS. I looked for the PHPMailer.php file on the server but I could not find it on the public_html/lib/pkp/lib/vendor/phpmailer/phpmailer/src/PHPMailer.php. This section of address "/src/PHPMailer.php" did not exist on the server! Is there any alternative solution to solve this problem?

@amirtkh , please take a look the Support Forum here:
https://forum.pkp.sfu.ca

You'll find a lot of existing guidance and debugging recommendations, but you can also open a new post for the community to help resolve your issue.
